### PR TITLE
[#24] initial commit.

### PR DESCRIPTION
**Pull request by tamerharsh**
**Implemented** 
1. Validity checker.
2. retrieve value as string.

**Please review.**

### CODE REVIEW

File: commandlineparser/include/Parser.h [hunk 1]
```diff
@@ -11,12 +10,13 @@
 #include <stdexcept>
 #include<tuple>
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: space <>

File: commandlineparser/include/Parser.h [hunk 2]
```diff
@@ -34,7 +34,7 @@ namespace cmdParser
 
 		//parses_the_input data and gives key value pair in  string type.
 		bool Parse(int argc,char* argv[]);
-
+		StorgeType ValueAsString;
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: why public ?  comment before the variables whats it is used for

File: commandlineparser/include/Parser.h [hunk 3]
```diff
@@ -54,11 +54,20 @@ namespace cmdParser
...
+		//to the store the corresponding value of input commands obtained by the commandline
+		void store_as_string(int argc, char**argv);
+
 		//a map for storing command Options.
 		CommandList command_list;
 		
 		//storing the tokenized data.		
 		std::vector<std::string> tokenized_data;
+
+		std::map<std::string, int> store_commands;
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: Please add comment

File: commandlineparser/src/Parser.cpp [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 void cmdParser::Parser::AddOptions(std::string short_command,std::string long_command,std::string short_description,std::string long_description) 
 {
-	std::shared_ptr<cmdParser::Options> obj = std::make_shared<cmdParser::Options>(short_command,long_command,short_description,long_description);
+    std::shared_ptr<cmdParser::Options> obj = std::make_shared<cmdParser::Options>(short_command,long_command,short_description,long_description);
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp:5]: Why this addition and removal of spaces.. every commit indentation is changed ? 
Indentation 
4 spaces
1 tab = 4 spaces

File: commandlineparser/src/Parser.cpp [hunk 5]
```diff
@@ -130,3 +133,77 @@ void cmdParser::Parser::long_help(const std::vector<std::string>&keys) const
...
+			if (command_list.find(temp) == command_list.end())
+			{
+				throw std::exception("Entered command not found in the registered command list");
+			}
+			
+			if (store_commands.find(temp) != store_commands.end())
+			{
+				store_commands.erase(temp);
+				//store_commands.insert({ temp, i });
+				store_commands.insert({ temp, i});
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp]: write a comment why deleteion and insertion are done

File: commandlineparser/src/Parser.cpp [hunk 6]
```diff
@@ -130,3 +133,77 @@ void cmdParser::Parser::long_help(const std::vector<std::string>&keys) const
...
+			{
+				store_commands.insert({ temp, i });
+			}
+		}
+	}
+}
+
+void cmdParser::Parser::store_as_string(int argc, char**argv)
+{
+	std::string key;
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp:177]: Just an idea.. may be in future you can write a overload function which uses regex for doing following logic

File: commandlineparser/include/Parser.h [hunk 7]
```diff
@@ -1,16 +1,15 @@
 #ifndef PARSER_H
 #define PARSER_H
 
+#include <stdio.h>
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: USE  IOSTREAM

File: commandlineparser/include/Parser.h [hunk 8]
```diff
@@ -34,7 +33,10 @@ namespace cmdParser
 
 		//parses_the_input data and gives key value pair in  string type.
 		bool Parse(int argc,char* argv[]);
-		StorgeType ValueAsString;
+
+		//to get the string value of query command.
+		std::vector<std::string> getValueAsString(const std::string &);
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: CONST

**Review by inbangsa — CHANGES_REQUESTED** [review]

File: commandlineparser/include/Parser.h (hunk 3 above)
Comment (tamerharsh) [reply commandlineparser/include/Parser.h]: done.

File: commandlineparser/src/Parser.cpp (hunk 5 above)
Comment (tamerharsh) [reply commandlineparser/src/Parser.cpp]: done.

File: commandlineparser/src/Parser.cpp (hunk 6 above)
Comment (tamerharsh) [reply commandlineparser/src/Parser.cpp:177]: ok.

File: commandlineparser/include/Parser.h (hunk 1 above)
Comment (tamerharsh) [reply commandlineparser/include/Parser.h]: removed.

File: commandlineparser/include/Parser.h (hunk 2 above)
Comment (tamerharsh) [reply commandlineparser/include/Parser.h]: corrected.

File: commandlineparser/src/Parser.cpp [hunk 9]
```diff
@@ -135,34 +142,33 @@ void cmdParser::Parser::long_help(const std::vector<std::string>&keys) const
...
+				else
+				{
+					store_commands.insert({ temp, i });
+				}
 		}
-	}
+	}	
 }
 
 void cmdParser::Parser::store_as_string(int argc, char**argv)
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp:175]: Combine this with validity checker

Comment (inbangsa) [reply commandlineparser/src/Parser.cpp:175]: Both are having same loops

File: commandlineparser/src/Parser.cpp [hunk 10]
```diff
@@ -135,34 +142,33 @@ void cmdParser::Parser::long_help(const std::vector<std::string>&keys) const
 }
 
 void cmdParser::Parser::validity_checker(int argc, char *argv[])
-{
+{   //to store only commands extracted from command line for  the validity check with the registered options.
+	std::map<std::string, int> store_commands;
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp:144]: what is the use of this map.. as this is local so the stored values  cannot be used outside.. and then what is the need to fill this map..

File: commandlineparser/src/Parser.cpp [hunk 11]
```diff
@@ -62,14 +61,22 @@ bool cmdParser::Parser::Parse(int argc, char* argv[])
...
 	};
 
-	call_help("-h", std::bind(&cmdParser::Parser::short_help, this, std::placeholders::_1));
-	call_help("--help", std::bind(&cmdParser::Parser::long_help, this, std::placeholders::_1));
+bool called_help_1=call_help("-h", std::bind(&cmdParser::Parser::short_help, this, std::placeholders::_1));
+bool called_help_2=call_help("--help", std::bind(&cmdParser::Parser::long_help, this, std::placeholders::_1));
 
-	validity_checker(argc, argv);
-	store_as_string(argc, argv);
+	if (called_help_1 == 0 && called_help_2 == 0)
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp]: try to remove this loops

**Review by inbangsa — CHANGES_REQUESTED** [review]

**Comment by inbangsa** [comment]
Not needed

**Pull request closed without merging**